pkg/registry/core/service/allocator/patch_bitmap_test.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package allocator
+
+import (
+	"testing"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+func TestAllocate_BitmapReserved(t *testing.T) {
+	max := 20
+	m := NewAllocationMapReserved(max, "test")
+
+	if _, ok, _ := m.AllocateNext(); !ok {
+		t.Fatalf("unexpected error")
+	}
+	if m.count != 1 {
+		t.Errorf("expect to get %d, but got %d", 1, m.count)
+	}
+	if f := m.Free(); f != max-1 {
+		t.Errorf("expect to get %d, but got %d", max-1, f)
+	}
+}
+
+// TestAllocateMaxReserved depends on the number of reserved offsets
+// Currently there is only one value reserved, if this value change
+// in the future the test has to be modified accordinly
+func TestAllocateMax_BitmapReserved(t *testing.T) {
+	max := 20
+	// modify if necessary
+	reserved := 1
+	m := NewAllocationMapReserved(max, "test")
+	for i := 0; i < max-reserved; i++ {
+		if _, ok, _ := m.AllocateNext(); !ok {
+			t.Fatalf("unexpected error")
+		}
+	}
+
+	if _, ok, _ := m.AllocateNext(); ok {
+		t.Errorf("unexpected success")
+	}
+	if f := m.Free(); f != reserved {
+		t.Errorf("expect to get %d, but got %d", 0, f)
+	}
+}
+
+func TestAllocateError_BitmapReserved(t *testing.T) {
+	m := NewAllocationMapReserved(20, "test")
+	if ok, _ := m.Allocate(3); !ok {
+		t.Errorf("error allocate offset %v", 3)
+	}
+	if ok, _ := m.Allocate(3); ok {
+		t.Errorf("unexpected success")
+	}
+}
+
+// 9 is a reserved value used for OpenshiftDNS
+// it can only be allocated explicitly using Allocate()
+func TestAllocateReservedOffset_BitmapReserved(t *testing.T) {
+	m := NewAllocationMapReserved(20, "test")
+	if ok, _ := m.Allocate(9); !ok {
+		t.Errorf("error allocate offset %v", 9)
```

Comment:
This should not actually be an `Errorf` log, but it could be `Infof`, correct?
```suggestion
		t.Infof("expected error allocating offset 9")
```

Reply:
it is an error, the 9 offset is the .10 address (the .1 address is the 0, because we don't count the subnet address), and it must be free on a recently created bitmap.
This tests verifies the explicitly allocation of the .10 address https://github.com/openshift/kubernetes/pull/921/files#r705722602
and checks that it can't be allocated twice

Reply:
It would be useful to have a test that mimics the behavior of the function that was obtaining the `.10` address before.  I think we might want to test what happens when all that's left is the `.10` address, but we still want to keep it reserved.

Reply:
I couldn't easily find code where the Allocate/AllocateNext might be used, but I'm just interested in some test of: https://github.com/openshift/kubernetes/pull/921/files#r706351153

Reply:
yeah, I'm missing that test case

Reply:
added